### PR TITLE
ci: pin nightly compiler, add CI cronjob to update it.

### PR DESCRIPTION
**Pull request by apoelstra**
We should be pinning our nightly compiler in CI so that
* running CI on old release branches will continue to work (not quite -- we'd want to do the same thing for Cargo.lock, but ok, one PR at a time)
* when CI breaks due to nightly updates (usually new Clippy lints), we can address that in a dedicated "update nightly" PR rather than having every other PR suddenly break on us

Co-authored with ChatGPT which found several typos and suggested most of the error-handling logic.

### CODE REVIEW

File: contrib/run_task.sh
Comment (tcharding) [inline contrib/run_task.sh:249]:
```suggestion
    if echo "$cargo_ver" | grep nightly; then
```

Comment (tcharding) [reply contrib/run_task.sh:249]: Also, shouldn't this be `grep -v`?

Comment (apoelstra) [reply contrib/run_task.sh]: Yes :) two mistakes cancelled themselves out.

Comment (tcharding) [reply contrib/run_task.sh]: `shellcheck` would have caught that I rekon.

**Review by tcharding — APPROVED** [review]
ACK 6ca0dd7b5e64683519b023589ccf5081a413f391

**Comment by tcharding** [comment]
Oh and woops on leaving the `contrib/tesh.sh` scripts in there.

File: contrib/run_task.sh
Comment (apoelstra) [reply contrib/run_task.sh]: `shellcheck` isn't able to check spelling nor can it tell whether I meant `-v` ... but I think ChatGPT would've caught this had I run the diff through there.

**Comment by apoelstra** [comment]
Will wait for @Kixunil to review. Arguably this is "CI only" and we can one-ACK merge, but it's fairly nontrivial and changes our CI pipeline.

**Comment by tcharding** [comment]
Likely a good idea, I did not put too much effort into reading the actual github action code.

**Comment by apoelstra** [comment]
The GA code I ran through ChatGPT repeatedly and we can't really test it until the cronjob triggers and we see what it does. My feeling is that it's ok if we merge it and it's broken and we have to iterate a bit. (Though I think there is a way I can trigger it manually, which we should try after merging....or before merging if it's possible.)

**Comment by tcharding** [comment]
> My feeling is that it's ok if we merge it and it's broken and we have to iterate a bit.

That was my feeling too.

**Comment by tcharding** [comment]
Can we one ack merge and @Kixunil comes back to post-merge-review later on. Broken CI is killing me.

**Comment by apoelstra** [comment]
Oh, oops, I *did* mean to one-ACK merge this before Sunday at 00:00 UTC, but it looks like I missed it.

@Kixunil can you take a look at this? Otherwise I'll merge it in the next few days.

File: .github/workflows/rust.yml
Comment (Kixunil) [inline .github/workflows/rust.yml:78]: Ironically this unpins other part of the CI: the action. If there's an unreleased feature needed (didn't check) then it'd be better to just use the commit hash.

File: .github/workflows/rust.yml
Comment (Kixunil) [inline .github/workflows/rust.yml:183]: Can we somehow make it into a variable? Should make it less annoying to keep jobs in sync.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (Kixunil) [inline .github/workflows/cron-weekly-update-nightly.yml:18]: Interesting trick, I always used `grep '^foo' | sed s/^foo: //`

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (Kixunil) [inline .github/workflows/cron-weekly-update-nightly.yml:23]: A variable would make this much simpler too.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (Kixunil) [inline .github/workflows/cron-weekly-update-nightly.yml:4]: I wanted to run a test job every day after nightly releases to help with uncovering bugs quickly but update the pin weekly sounds good.

**Review by Kixunil — APPROVED** [review]
ACK 6ca0dd7b5e64683519b023589ccf5081a413f391

This is better than the previous situation but please consider my comments.

File: .github/workflows/rust.yml
Comment (apoelstra) [reply .github/workflows/rust.yml:78]: I *believe* the latest nightlies come with master. But let me try pinning to some speciifc commit and see if it works.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (apoelstra) [reply .github/workflows/cron-weekly-update-nightly.yml:18]: Heh, I copied this out of dtolnay/rust-toolchain, since I had to dig through that repo for the syntax anyway.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (apoelstra) [reply .github/workflows/cron-weekly-update-nightly.yml:4]: I'll make this daily and we'll see how annoying it is. In theory it's a one-ACK merge from me or you or @tcharding and 98% of the time no issues will appear.

File: .github/workflows/rust.yml
Comment (apoelstra) [reply .github/workflows/rust.yml:183]: Yeah, lemme iterate with the chatbot.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (apoelstra) [reply .github/workflows/cron-weekly-update-nightly.yml:23]: In what way? No matter what I need to pull the opaque "nightly" compiler out of dtolnay/rust-toolchain and then query it to learn what its actual version is.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (yancyribbens) [inline .github/workflows/cron-weekly-update-nightly.yml:17]: Is this putting a lot of trust in what comes downstream from dtolnay?

File: CONTRIBUTING.md
Comment (yancyribbens) [inline CONTRIBUTING.md:127]: guess this isn't useful anymore?

File: .github/workflows/rust.yml
Comment (yancyribbens) [reply .github/workflows/rust.yml:183]: Can we call the chatbot HAL 9000? :)

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (apoelstra) [reply .github/workflows/cron-weekly-update-nightly.yml:18]: Resolving as this is non-actionable.

File: .github/workflows/rust.yml
Comment (apoelstra) [reply .github/workflows/rust.yml]: Ok, pinned to the `v1` tag which should be fine.

File: .github/workflows/rust.yml
Comment (apoelstra) [reply .github/workflows/rust.yml]: Added a file to a repo and read it from there.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (apoelstra) [reply .github/workflows/cron-weekly-update-nightly.yml]: Done. It is a bit simpler now in that we don't need to invoke `sed`.

File: .github/workflows/cron-weekly-update-nightly.yml
Comment (apoelstra) [reply .github/workflows/cron-weekly-update-nightly.yml]: It's the same amount of trust as in our existing CI scripts. (Which isn't much -- CI can't affect the repo.)

File: CONTRIBUTING.md
Comment (apoelstra) [reply CONTRIBUTING.md:127]: No, because it is out of sync with what's actually run in CI (it already does not include the `base58` crate for example)

**Comment by apoelstra** [comment]
@Kixunil updated to address your comments.

File: CONTRIBUTING.md
Comment (Kixunil) [reply CONTRIBUTING.md:127]: The individual scripts in each crate are out of date but the top one is up to date.

File: contrib/run_task.sh
Comment (Kixunil) [inline contrib/run_task.sh:249]: `-q` to not pollute the output would be nice.

File: .github/workflows/rust.yml
Comment (Kixunil) [inline .github/workflows/rust.yml:22]: You need to append this to `$GITHUB_OUTPUT` (with `>>`). Since not doing so is deprecated due to security problems.

**Review by Kixunil — COMMENTED** [review]
Really wanted to ACK this but the redirect issue seems quite problematic.

File: .github/workflows/rust.yml
Comment (apoelstra) [reply .github/workflows/rust.yml:22]: Oh, good to know, thanks!

**Comment by apoelstra** [comment]
@Kixunil addressed both comments.

Thanks for noticing so quickly when I finally got CI to pass!

**Review by Kixunil — DISMISSED** [review]
ACK 16c8aae5d30d13b7f122a724824726eb802f398e

File: .github/workflows/rust.yml
Comment (Kixunil) [inline .github/workflows/rust.yml:22]: Oh, sorry for dumb advice, the syntax also changed, so it should be

```sh
echo "nightly_version=$(cat .github/nightly-version)" >> $GITHUB_OUTPUT
```

File: contrib/run_task.sh
Comment (tcharding) [reply contrib/run_task.sh]: Oh of course, i was thinking nighly was an unset variable - face palm. 

> but I think ChatGPT would've caught this had I run the diff through there.

So you just paste a diff and ask "check this" and it gives you useful feedback? What version?

File: contrib/run_task.sh
Comment (tcharding) [inline contrib/run_task.sh:152]: Ah, sorry for not noticing last time. This line should be below the next one (the local var setting), right? Its splitting the comment above from the line it comments (I know this is shell but the old C programmer in me hurts a little bit to see a function call before the varible declaration :)

Can be done as a follow up though so we can get this in.

**Review by tcharding — APPROVED** [review]
ACK 651f8d9bcb9cc3248d5eb0dd92c446de594957b9

File: contrib/run_task.sh
Comment (tcharding) [reply contrib/run_task.sh:152]: Better context:
```diff
     # Use the current (recent/minimal) lock file.
-    local cargo="cargo +nightly --locked"
+    need_nightly
+    local cargo="cargo --locked"
```

Comment (apoelstra) [reply contrib/run_task.sh:152]: Yeah, sure, I'll move the comment up a line.

**Review by tcharding — APPROVED** [review]
ACK ee113aa91ffed1a296b112133685b45bde546326

**Review by Kixunil — APPROVED** [review]
ACK ee113aa91ffed1a296b112133685b45bde546326

**Comment by apoelstra** [comment]
@Kixunil @tcharding please resolve your open comments.

**Comment by tcharding** [comment]
I've done mine, thanks.

**Comment by apoelstra** [comment]
It's merged. I don't know why Github is failing to display this.

**Pull request closed without merging**
